pkg/lib/lib_test.go

```diff
@@ -770,28 +771,286 @@ func TestISO8583MessageWithValidData(t *testing.T) {
+	assert.Nil(t, err)
+
+	err = message.Validate()
 	assert.NotNil(t, err)
+
+	buf, err = message.Bytes()
+	assert.Nil(t, err)
+	assert.Equal(t, buf, byteData)
 }
 
-func TestISO8583MessageWithJson(t *testing.T) {
+func TestISO8583MessageWithBcdNumberEncoding(t *testing.T) {
+	jsonData, err := ioutil.ReadFile(filepath.Join("..", "..", "test", "testdata", "attributes_data_elements.dat"))
+	assert.Nil(t, err)
+
+	encoding := &utils.EncodingDefinition{
+		MtiEnc:       utils.EncodingChar,
+		BitmapEnc:    utils.EncodingHex,
+		LengthEnc:    utils.EncodingRBcd,
+		NumberEnc:    utils.EncodingBcd,
+		CharacterEnc: utils.EncodingAscii,
+		BinaryEnc:    utils.EncodingHex,
+		TrackEnc:     utils.EncodingEbcdic,
+	}
+	message, err := NewISO8583MessageWithAttributes(jsonData, encoding)
+	assert.Nil(t, err)
+
+	byteData, err := ioutil.ReadFile(filepath.Join("..", "..", "test", "testdata", "message_with_number_bcd_encoding.dat"))
+	assert.Nil(t, err)
+
+	_, err = message.Load(byteData)
+	assert.Nil(t, err)
+
+	err = message.Validate()
+	assert.Nil(t, err)
+
+	buf, err := message.Bytes()
+	assert.Nil(t, err)
+	assert.Equal(t, buf, byteData)
+}
+
+func TestISO8583MessageWithRBcdNumberEncoding(t *testing.T) {
+	jsonData, err := ioutil.ReadFile(filepath.Join("..", "..", "test", "testdata", "attributes_data_elements.dat"))
+	assert.Nil(t, err)
+
+	encoding := &utils.EncodingDefinition{
+		MtiEnc:       utils.EncodingChar,
+		BitmapEnc:    utils.EncodingHex,
+		LengthEnc:    utils.EncodingRBcd,
+		NumberEnc:    utils.EncodingRBcd,
+		CharacterEnc: utils.EncodingAscii,
+		BinaryEnc:    utils.EncodingHex,
+		TrackEnc:     utils.EncodingEbcdic,
+	}
+	message, err := NewISO8583MessageWithAttributes(jsonData, encoding)
+	assert.Nil(t, err)
+
+	byteData, err := ioutil.ReadFile(filepath.Join("..", "..", "test", "testdata", "message_with_number_bcd_encoding.dat"))
+	assert.Nil(t, err)
+
+	_, err = message.Load(byteData)
+	assert.Nil(t, err)
+
+	err = message.Validate()
+	assert.Nil(t, err)
+
+	buf, err := message.Bytes()
+	assert.Nil(t, err)
+	assert.Equal(t, buf, byteData)
+}
+
+func TestISO8583MessageWithCharacterEbcdicEncoding(t *testing.T) {
 	jsonData, err := ioutil.ReadFile(filepath.Join("..", "..", "test", "testdata", "attributes_data_elements.dat"))
 	assert.Nil(t, err)
 
-	_, err = NewISO8583MessageWithJson(jsonData, nil)
+	encoding := &utils.EncodingDefinition{
+		MtiEnc:       utils.EncodingChar,
+		BitmapEnc:    utils.EncodingHex,
+		LengthEnc:    utils.EncodingChar,
+		NumberEnc:    utils.EncodingChar,
+		CharacterEnc: utils.EncodingEbcdic,
+		BinaryEnc:    utils.EncodingHex,
+		TrackEnc:     utils.EncodingEbcdic,
+	}
+	message, err := NewISO8583MessageWithAttributes(jsonData, encoding)
+	assert.Nil(t, err)
+
+	byteData, err := ioutil.ReadFile(filepath.Join("..", "..", "test", "testdata", "message_with_character_ebcdic_encoding.dat"))
+	assert.Nil(t, err)
+
+	_, err = message.Load(byteData)
+	assert.Nil(t, err)
+
+	err = message.Validate()
```

Comment:
After we validate the message, I would expect us to test that we have valid/expected values in the specific fields. Like:

```go
elements := message.GetElements()
assert.Equal(t, "xxxx", elements[4].String())
``` 

Maybe not for all .dat files, but at least for one.

Reply:
fixed